File: variant-a/java-service/src/main/java/com/flashsale/api/service/AllocationManagerV2.java
package com.flashsale.api.service;

import com.flashsale.api.entity.CampaignSkuAllocation;
import com.flashsale.api.repository.CampaignSkuAllocationRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.net.InetAddress;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Allocation Manager V2 - Claims and manages allocation units for service instances
 *
 * Responsibilities:
 * - Claim allocation units from database on startup
 * - Load claimed units into AdaptiveInventoryV2 managers
 * - Provide access to adaptive inventory instances
 */
@Service
public class AllocationManagerV2 {

    private static final Logger logger = LoggerFactory.getLogger(AllocationManagerV2.class);

    private final CampaignSkuAllocationRepository allocationRepository;
    private final RedisTemplate<String, Object> redisTemplate;

    // Service instance ID (unique per container/pod)
    private final String serviceId;

    // Map of SKU ID → AdaptiveInventoryV2 instance
    private final Map<UUID, AdaptiveInventoryV2> inventoryManagers = new ConcurrentHashMap<>();

    // Track claimed allocation IDs
    private final List<Long> claimedAllocationIds = new ArrayList<>();

    public AllocationManagerV2(
            CampaignSkuAllocationRepository allocationRepository,
            RedisTemplate<String, Object> redisTemplate) {

        this.allocationRepository = allocationRepository;
        this.redisTemplate = redisTemplate;

        // Generate unique service ID (hostname + UUID)
        this.serviceId = generateServiceId();
    }

    /**
     * Generate unique service ID for this instance
     * Format: java-{hostname}-{uuid}
     */
    private String generateServiceId() {
        try {
            String hostname = InetAddress.getLocalHost().getHostName();
            String shortUuid = UUID.randomUUID().toString().substring(0, 12);
            return String.format("java-%s-%s", hostname, shortUuid);
        } catch (Exception e) {
            logger.warn("Failed to get hostname, using UUID only", e);
            return String.format("java-%s", UUID.randomUUID().toString().substring(0, 12));
        }
    }

    /**
     * Claim allocation units and load into memory
     *
     * @param campaignId Campaign to claim units for
     * @return Number of units claimed
     */
    @Transactional
    public int claimAndLoadAllocations(UUID campaignId) {
        logger.info("=" + "=".repeat(60));
        logger.info("VARIANT A: Initializing Adaptive Inventory System");
        logger.info("=" + "=".repeat(60));
        logger.info("Service ID: {}", serviceId);

        // Determine max units to claim based on service capacity
        int maxUnits = getMaxUnitsForService();

        // Claim units atomically
        int claimedCount = allocationRepository.claimUnits(
            serviceId,
            campaignId.toString(),
            maxUnits
        );

        if (claimedCount == 0) {
            logger.warn(
                "⚠️  No allocation units claimed! " +
                "This service will fall back to Redis campaign pool immediately."
            );
            return 0;
        }

        logger.info("Claimed {} allocation units for campaign {}", claimedCount, campaignId);

        // Load claimed units into memory
        List<CampaignSkuAllocation> units = allocationRepository.findByClaimedByAndCampaignId(
            serviceId,
            campaignId
        );

        int totalAllocated = 0;
        for (CampaignSkuAllocation unit : units) {
            // Create AdaptiveInventoryV2 instance
            AdaptiveInventoryV2 inventory = new AdaptiveInventoryV2(
                unit.getId(),
                unit.getCampaignId(),
                unit.getSkuId(),
                unit.getAllocatedQuantity(),
                unit.getRefillBatchSize(),
                unit.getLowWaterMarkPct(),
                redisTemplate
            );

            // Store by SKU ID (for quick lookup during order processing)
            inventoryManagers.put(unit.getSkuId(), inventory);
            claimedAllocationIds.add(unit.getId());

            totalAllocated += unit.getAllocatedQuantity();

            logger.info(
                "Loaded allocation unit {}: SKU={}, qty={}, batch={}, watermark={}%",
                unit.getId(), unit.getSkuId(), unit.getAllocatedQuantity(),
                unit.getRefillBatchSize(), unit.getLowWaterMarkPct()
            );

            // Log initial metrics
            AdaptiveInventoryV2.InventoryMetrics metrics = inventory.getMetrics();
            logger.info(
                "  Unit {}: {} items in RAM, low_water_mark={}",
                metrics.allocationId, metrics.localStock, metrics.currentLowWaterMark
            );
        }

        logger.info("Total inventory loaded: {} items across {} units", totalAllocated, claimedCount);
        logger.info("=" + "=".repeat(60));
        logger.info("✓ Adaptive Inventory Initialization Complete");
        logger.info("  Ready to serve requests from local memory!");
        logger.info("=" + "=".repeat(60));

        return claimedCount;
    }

    /**
     * Get adaptive inventory manager for a SKU
     *
     * @param skuId SKU ID
     * @return AdaptiveInventoryV2 instance, or null if not found
     */
    public AdaptiveInventoryV2 getInventoryManager(UUID skuId) {
        return inventoryManagers.get(skuId);
    }

    /**
     * Check if service has allocation units for any SKU
     */
    public boolean hasAllocations() {
        return !inventoryManagers.isEmpty();
    }

    /**
     * Get all metrics from claimed allocation units
     */
    public Map<String, Object> getAllMetrics() {
        Map<String, Object> allMetrics = new HashMap<>();

        List<Map<String, Object>> unitMetrics = new ArrayList<>();
        for (AdaptiveInventoryV2 inventory : inventoryManagers.values()) {
            AdaptiveInventoryV2.InventoryMetrics m = inventory.getMetrics();

            Map<String, Object> metrics = new HashMap<>();
            metrics.put("allocation_id", m.allocationId);
            metrics.put("local_stock", m.localStock);
            metrics.put("current_low_water_mark", m.currentLowWaterMark);
            metrics.put("total_requests", m.totalRequests);
            metrics.put("ram_hits", m.ramHits);
            metrics.put("ram_hit_rate_pct", String.format("%.2f", m.getRamHitRate()));
            metrics.put("redis_refills", m.redisRefills);
            metrics.put("campaign_pool_hits", m.campaignPoolHits);
            metrics.put("refill_in_progress", m.refillInProgress);

            unitMetrics.add(metrics);
        }

        allMetrics.put("service_id", serviceId);
        allMetrics.put("claimed_units", claimedAllocationIds.size());
        allMetrics.put("unit_metrics", unitMetrics);

        return allMetrics;
    }

    /**
     * Calculate max units to claim based on service capacity
     * Can be configured via environment variables or service discovery
     *
     * @return Maximum number of units to claim
     */
    private int getMaxUnitsForService() {
        // For now, use simple heuristic: 2 units per service instance
        // In production, this could be based on:
        // - Available memory
        // - CPU cores
        // - Service tier/capacity
        // - Environment variable configuration

        String maxUnitsEnv = System.getenv("MAX_ALLOCATION_UNITS");
        if (maxUnitsEnv != null) {
            try {
                return Integer.parseInt(maxUnitsEnv);
            } catch (NumberFormatException e) {
                logger.warn("Invalid MAX_ALLOCATION_UNITS value: {}", maxUnitsEnv);
            }
        }

        return 2; // Default: 2 units per service
    }

    public String getServiceId() {
        return serviceId;
    }

    public List<Long> getClaimedAllocationIds() {
        return new ArrayList<>(claimedAllocationIds);
    }
}

File: variant-a/csharp-service/Services/AllocationManagerV2.cs
using FlashSale.Api.Models;
using StackExchange.Redis;
using System.Collections.Concurrent;
using System.Net;

namespace FlashSale.Api.Services;

/// <summary>
/// Allocation Manager V2 - Claims and manages allocation units for service instances
///
/// Responsibilities:
/// - Claim allocation units from database on startup
/// - Load claimed units into AdaptiveInventoryV2 managers
/// - Provide access to adaptive inventory instances
/// </summary>
public class AllocationManagerV2
{
    private readonly ILogger<AllocationManagerV2> _logger;
    private readonly ICampaignSkuAllocationRepository _allocationRepository;
    private readonly IDatabase _redis;
    private readonly IServiceProvider _serviceProvider;

    // Service instance ID (unique per container/pod)
    private readonly string _serviceId;

    // Map of SKU ID → AdaptiveInventoryV2 instance
    private readonly ConcurrentDictionary<Guid, AdaptiveInventoryV2> _inventoryManagers = new();

    // Track claimed allocation IDs
    private readonly List<long> _claimedAllocationIds = new();

    public AllocationManagerV2(
        ILogger<AllocationManagerV2> logger,
        ICampaignSkuAllocationRepository allocationRepository,
        IConnectionMultiplexer redis,
        IServiceProvider serviceProvider)
    {
        _logger = logger;
        _allocationRepository = allocationRepository;
        _redis = redis.GetDatabase();
        _serviceProvider = serviceProvider;

        // Generate unique service ID (hostname + UUID)
        _serviceId = GenerateServiceId();
    }

    /// <summary>
    /// Generate unique service ID for this instance
    /// Format: csharp-{hostname}-{uuid}
    /// </summary>
    private string GenerateServiceId()
    {
        try
        {
            string hostname = Dns.GetHostName();
            string shortUuid = Guid.NewGuid().ToString()[..12];
            return $"csharp-{hostname}-{shortUuid}";
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "Failed to get hostname, using UUID only");
            return $"csharp-{Guid.NewGuid().ToString()[..12]}";
        }
    }

    /// <summary>
    /// Claim allocation units and load into memory
    /// </summary>
    public async Task<int> ClaimAndLoadAllocationsAsync(Guid campaignId)
    {
        _logger.LogInformation("============================================================");
        _logger.LogInformation("VARIANT A: Initializing Adaptive Inventory System");
        _logger.LogInformation("============================================================");
        _logger.LogInformation("Service ID: {ServiceId}", _serviceId);

        // Determine max units to claim based on service capacity
        int maxUnits = GetMaxUnitsForService();
        _logger.LogInformation("Attempting to claim up to {MaxUnits} allocation units...", maxUnits);

        // Claim units atomically from database
        int claimedCount = await _allocationRepository.ClaimUnitsAsync(
            _serviceId,
            campaignId.ToString(),
            maxUnits
        );

        if (claimedCount == 0)
        {
            _logger.LogWarning("No allocation units available to claim for campaign {CampaignId}", campaignId);
            return 0;
        }

        _logger.LogInformation("Successfully claimed {ClaimedCount} allocation units", claimedCount);

        // Load claimed units into memory
        var claimedUnits = await _allocationRepository.GetClaimedUnitsAsync(
            _serviceId,
            campaignId.ToString()
        );

        _logger.LogInformation("Loading {Count} allocation units into RAM...", claimedUnits.Count);

        foreach (var unit in claimedUnits)
        {
            try
            {
                // Create logger for this specific inventory instance
                var inventoryLogger = _serviceProvider.GetRequiredService<ILogger<AdaptiveInventoryV2>>();

                // Create AdaptiveInventoryV2 instance
                var inventory = new AdaptiveInventoryV2(
                    unit.Id,
                    Guid.Parse(unit.CampaignId),
                    Guid.Parse(unit.SkuId),
                    unit.AllocatedQuantity,
                    unit.RefillBatchSize,
                    unit.LowWaterMarkPct,
                    _redis,
                    inventoryLogger
                );

                // Store in map (SKU ID → inventory)
                _inventoryManagers[Guid.Parse(unit.SkuId)] = inventory;
                _claimedAllocationIds.Add(unit.Id);

                _logger.LogInformation(
                    "  Loaded allocation unit {AllocationId}: SKU={SkuId}, qty={Qty}, refill={Refill}, low_water={LowWater}%",
                    unit.Id, unit.SkuId, unit.AllocatedQuantity, unit.RefillBatchSize, unit.LowWaterMarkPct * 100
                );
            }
            catch (Exception ex)
            {
                _logger.LogError(ex,
                    "Failed to load allocation unit {AllocationId}", unit.Id);
            }
        }

        _logger.LogInformation("============================================================");
        _logger.LogInformation("Successfully claimed and loaded {Count} allocation units into RAM", claimedUnits.Count);
        _logger.LogInformation("Total items in RAM: {TotalItems}",
            claimedUnits.Sum(u => u.AllocatedQuantity));
        _logger.LogInformation("============================================================");

        return claimedCount;
    }

    /// <summary>
    /// Get adaptive inventory manager for a SKU
    /// </summary>
    public AdaptiveInventoryV2? GetInventoryForSku(Guid skuId)
    {
        _inventoryManagers.TryGetValue(skuId, out var inventory);
        return inventory;
    }

    /// <summary>
    /// Check if this service has claimed units for a SKU
    /// </summary>
    public bool HasInventoryForSku(Guid skuId)
    {
        return _inventoryManagers.ContainsKey(skuId);
    }

    /// <summary>
    /// Get all metrics for monitoring
    /// </summary>
    public Dictionary<Guid, AdaptiveInventoryMetrics> GetAllMetrics()
    {
        var metrics = new Dictionary<Guid, AdaptiveInventoryMetrics>();

        foreach (var entry in _inventoryManagers)
        {
            metrics[entry.Key] = entry.Value.GetMetrics();
        }

        return metrics;
    }

    /// <summary>
    /// Determine max units to claim based on service capacity
    /// Can be configured via environment variable or use defaults
    /// </summary>
    private int GetMaxUnitsForService()
    {
        // Try to read from environment variable
        string? maxUnitsEnv = Environment.GetEnvironmentVariable("MAX_ALLOCATION_UNITS");
        if (!string.IsNullOrEmpty(maxUnitsEnv) && int.TryParse(maxUnitsEnv, out int maxUnits))
        {
            return maxUnits;
        }

        // Default: claim 2-4 units based on available CPU cores
        int cpuCores = Environment.ProcessorCount;
        if (cpuCores >= 8)
        {
            return 4;  // High-capacity server
        }
        else if (cpuCores >= 4)
        {
            return 2;  // Medium-capacity server
        }
        else
        {
            return 2;  // Low-capacity server
        }
    }

    /// <summary>
    /// Release all claimed allocation units (for graceful shutdown)
    /// </summary>
    public async Task ReleaseAllAllocationsAsync()
    {
        _logger.LogInformation("Releasing all claimed allocation units for service {ServiceId}...", _serviceId);

        await _allocationRepository.ReleaseUnitsAsync(_serviceId);

        _inventoryManagers.Clear();
        _claimedAllocationIds.Clear();

        _logger.LogInformation("All allocation units released successfully");
    }
}
